### PR TITLE
Add CorrCholesky transform

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR adds CorrCholesky transform. Because each row of the triangular part of the cholesky of a correlation matrix has norm 1 (or diagonal part of correlation matrix is 1), we can apply StickBreaking process for the square of each row. Using that fact, we can vectorize this transform and avoid "for" loops in the transform.

### CODE REVIEW

File: test/test_distributions.py [hunk 1]
```diff
@@ -434,8 +439,9 @@ def test_constraints(constraint, x, expected):
     assert_array_equal(constraint(x), expected)
 
 
-@pytest.mark.parametrize('shape', [(), (1,), (3,), (5,), (3, 1), (1, 3), (5, 3)], ids=idfn)
+@pytest.mark.parametrize('shape', [(), (1,), (3,), (6,), (3, 1), (1, 3), (5, 3)], ids=idfn)
```
Comment (fehiepsi) [inline test/test_distributions.py]: We change this to 6 to make sure that it is the length of a lower tril vector (of 4 x 4 matrix)

File: test/test_distributions.py [hunk 2]
```diff
@@ -467,8 +473,14 @@ def test_biject_to(constraint, shape):
     assert np.shape(actual) == batch_shape
     if len(shape) == transform.event_dim:
         if constraint is constraints.simplex:
-            expected = np.linalg.slogdet(jax.jacobian(transform)(x)[:-1, :])[1]
-            inv_expected = np.linalg.slogdet(jax.jacobian(transform.inv)(y)[:, :-1])[1]
+            expected = onp.linalg.slogdet(jax.jacobian(transform)(x)[:-1, :])[1]
```
Comment (fehiepsi) [inline test/test_distributions.py]: using `onp` because `slogdet` does not work in GPU

**Comment by neerajprad** [comment]
Looks great! I will take a detailed look later tonight and merge this.

File: test/test_distributions.py (hunk 2 above)
Comment (neerajprad) [reply test/test_distributions.py]: How is this handled on the gpu? Does jax send the array to the CPU for the numpy operation and transfer the result back?

File: numpyro/distributions/transforms.py [hunk 3]
```diff
@@ -129,6 +129,91 @@ def log_abs_det_jacobian(self, x, y):
...
+        # inverse stick-breaking
+        z1m_cumprod = 1 - cumsum(y * y)
+        pad_width = [(0, 0)] * y.ndim
+        pad_width[-1] = (1, 0)
+        z1m_cumprod_shifted = np.pad(z1m_cumprod[..., :-1], pad_width,
+                                     mode="constant", constant_values=1.)
+        t = matrix_to_tril_vec(y, diagonal=-1) / np.sqrt(
+            matrix_to_tril_vec(z1m_cumprod_shifted, diagonal=-1))
+        # inverse of tanh
+        x = np.log((1 + t) / (1 - t)) / 2
```
Comment (neerajprad) [inline numpyro/distributions/transforms.py]: Can we test how numerically stable this is, e.g. by applying the forward and inverse transforms? The LKJ implementation in Pyro brought up issues with numerical stability, but I do not recall what they were.

Comment (fehiepsi) [reply numpyro/distributions/transforms.py]: We did this test for all transforms at [this line](https://github.com/pyro-ppl/numpyro/pull/126/files#diff-51c2d9ee9d114f18441e3579ec31fdbeR466).

Comment (fehiepsi) [reply numpyro/distributions/transforms.py]: For some reason, I can't reply to the `slogdet` comment. In GPU, `onp.linalg.slogdet` returns a numpy array.

File: test/test_distributions.py (hunk 2 above)
Comment (fehiepsi) [reply test/test_distributions.py]: In GPU, `onp.linalg.slogdet` returns a numpy array.

Comment (neerajprad) [reply test/test_distributions.py]: Wouldn't that be an issue when we try to use that with a GPU based array?

**Pull request merged**